### ISSUE TITLE
What is the Advantage of REX-Ray for S3FS?

### ISSUE THREAD

**Issue by geekanin** [rexray/rexray#1286]
I have been prototyping using S3FS as persistent storage for apps running in Docker containers. I have prototypes the use S3FS without REX-Ray and that use the REX-Ray S3FS Storage Driver, and so far am not understanding, for this specific use case, what are the advantages of using the REX-Ray driver. I am connecting to an on prem S3 target and not AWS S3.

For the S3FS standalone I mount the bucket to a local folder with S3FS, and pass the folder to the container as a bind mount:

```
# s3fs bucket-1 /mnt/s3fs/bucket-1/ -o url=https://<s3-endpoint>/ -o no_check_certificate -o passwd_file=/.passwd-s3fs 
# docker run -p 80:80 -v /mnt/s3fs/bucket-1/:/s3-vol my-app
```

For the REX-Ray Example
```
# docker plugin install rexray/s3fs S3FS_ENDPOINT="http://<s3-endpoint>" S3FS_OPTIONS="url=http://<s3-endpoint>" S3FS_ACCESSKEY=<access-key-id> S3FS_SECRETKEY=<secret-access-key> 
# docker -D -l debug volume create --driver rexray/s3fs:latest myrexvol-1
# docker run -p 80:80 --mount source=myrexvol-1,target=/s3-vol friendlyhello
```
You can see above I use http because https as shown below does not work with REX-Ray S3FS and self signed certs, despite adding the _no_check_certificate_ option
```
# docker plugin install rexray/s3fs S3FS_ENDPOINT="https://<s3-endpoint>" S3FS_OPTIONS="no_check_certificate,url=https://<s3-endpoint>" S3FS_ACCESSKEY=<access-key-id> S3FS_SECRETKEY=<secret-access-key> 
```

What I am trying to understand is why I would use REX-Ray for container access to S3FS. Of the two examples above, the S3FS only example is less complex than the REX-Ray example, and it works with self signed certs where REX-Ray does not appear to. What I am unhappy about in both examples is they both require a fair amount of host level dependency. In both cases you need to prepare the S3FS volumes on the Host before you start the containers. 

I was hoping that REX-Ray would give me a portable solution where the S3FS volumes might be defined in my Docker service docker-compose.yml file, and the mounting of those volumes would happen at runtime. In which case I could run my apps on any Docker host without prerequisite (i.e. portable).  Or I would be happy with only having to configure the REX-Ray S3FS volume on one or two hosts (libstorage servers) and connect my containers (libstorage clients) to those hosts at runtime, presenting the volumes as a local mount to my container application. Unfortunately I cannot figure out how to achieve that and am wondering if it is even possible.

I apologize if this is not the correct forum, I see other questions posted here so am hoping this is OK. Can anyone point me in the right direction? What am I missing?

Thanks!

**Comment by jnj16180340** [comment]
Given that s3fs-fuse is broken on Alpine linux/libc-musl AKA the docker plugin environment, my personal experience has been that the docker plugin (on AWS linux anyway) is basically a non-starter right now.

**Comment by geekanin** [comment]
> Given that s3fs-fuse is broken on Alpine linux/libc-musl AKA the docker plugin environment, my personal experience has been that the docker plugin (on AWS linux anyway) is basically a non-starter right now.

Is there a particular bug you are referring to? It is not currently my intent to run this in AWS EC2 but rather as part of an on prem solution. Thanks for the reply.

**Comment by jnj16180340** [comment]
Yes, search issues here+ in s3fs-fuse for "musl" "Alpine" or similar. I'm
currently working through an ECS deployment

**Comment by RothAndrew** [comment]
Services like AWS Elastic Container Service (ECS) support Docker Plugins. rexray/s3fs is a Docker Plugin, s3fs standalone is not.
